xla/service/while_loop_all_reduce_code_motion.h

```diff
@@ -58,6 +60,10 @@ class WhileLoopAllReduceCodeMotion : public HloModulePass {
 
  private:
   const bool enable_reduce_scatter_;
+
+  // Whether to run passes that may setup the add(all-reduce/reduce-scatter,
```

Comment:
Can you elaborate? It is not clear to me what this does

Reply:
So WhileLoopAllReduceCodeMotion can code motion data parallel all-reduce and reduce-scatters out of while loop and into the matin body. It can only code motion one specific pattern which is 
```
add(all-reduce/reduce-scatter, get-tuple-element(while_loop_param))
```
Note it also handle one more pattern involving dynamic-update-slice found in RepeatedTransformer. 

What this PR does is it allows handling convert and transpose ops. Convert is very heavily found in any Jax model using mixed precision training and transpose is found in models using a FusedQKV weight like so (3, hidden, heads, hidden_per_head.